xarray/core/groupby.py

```diff
@@ -379,7 +401,9 @@ def apply(self, func, shortcut=False, **kwargs):
             grouped = self._iter_grouped_shortcut()
         else:
             grouped = self._iter_grouped()
-        applied = (maybe_wrap_array(arr, func(arr, **kwargs)) for arr in grouped)
+        applied = (self._maybe_unstack_array(
+                    maybe_wrap_array(arr,func(arr, **kwargs)))
+                    for arr in grouped)
         combined = self._concat(applied, shortcut=shortcut)
         result = self._maybe_restore_empty_groups(combined)
```

Comment:
Yes, that seems obvious now! (But only in retrospect. ;)

I just tried putting the call to ``_maybe_unstack_arrays`` around `combined`.

But the problem is that the multi-index is no longer present in `combined`. If I add a call to `print (combined.indexes)`, I get

```
stacked_ny_nx: Index([(0, 0), (0, 1), (1, 0), (1, 1)], dtype='object', name=u'stacked_ny_nx')
```

i.e. just a regular index. This makes unstacking impossible (or a lot harder). I guess this is related to #769.

Unfortunately I have to copy the group variable because of the pandas / cython bug referenced above.

So kind of stuck.